### PR TITLE
Filter developers on mobile

### LINKED ISSUE #293: Filter /developers on mobile
Currently, the filters on `/developers` are only visible on desktop and tablets. A PR addressing this issue should add the same filters to the mobile version of the site.

I don't think we need a new Stimulus controller - a couple of nested `ToggleController.js` controllers should do the trick.

---

The existing filtering design was based off of the [With centered text and dropdown product filters](https://tailwindui.com/components/ecommerce/components/category-filters#component-cb75aacd7c636865835476c98f606a38) from Tailwind UI. Here are the comparisons of desktop to mobile.

image

image

image

Please only work on this issue if you have access to Tailwind UI. I want to keep the markup and classes the same so changes in the future are easier.

If you _really_ want to work on this but don't have access to Tailwind UI then [email me](mailto:joe@masilotti.com) and we can work something out!

### PR DESCRIPTION
**Pull request by dhairyagabha**
Fixes #293 and #305 

### Pull request checklist

- [x] Your code contains tests relevant for code you modified
- [x] You have linted and tested the project with `bin/check`

Screen Shot 2022-02-27 at 7 03 31 PM

Mobile Filters:
Screen Shot 2022-02-27 at 7 04 26 PM

### CODE REVIEW

**Comment by joemasilotti** [comment]
Is this ready for review? I see a TODO comment and didn't notice anywhere where the chevron is getting rotated. Unless I missed it!

**Comment by dhairyagabha** [comment]
My bad. Need to remove the TODO comment. Chevron is not rotated because it will require custom javascript just to rotate the chevron. Wondering if it is required enough to just write that.

**Comment by dhairyagabha** [comment]
@joemasilotti Its been updated with the chevron rotations as well.

File: app/components/developer_query_component.html.erb [hunk 1]
```diff
@@ -83,8 +83,88 @@
...
+                </div>
+              </div>
+            <% end %>
+          </div>
         </div>
       </div>
-    <% end %>
+    </div>
+    </div>
+    </div>
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:168]: It looks like this file needs to be formatted for consistent indentation.

File: app/components/developer_query_component.html.erb [hunk 2]
```diff
@@ -83,8 +83,88 @@
...
+        <button type="button" data-action="toggle#toggle" class="group inline-flex items-center justify-center text-sm font-medium text-gray-700 hover:text-gray-900" aria-expanded="false">
+          <span><%= t(".filters") %></span>
+        </button>
+      </div>
+      <div data-toggle-target="element" class="bg-gray-50">
+        <div class="fixed inset-0 flex z-40 sm:hidden" role="dialog" aria-modal="true">
+          <div class="fixed inset-0 bg-black bg-opacity-25" aria-hidden="true"></div>
+          <div class="ml-auto relative max-w-xs w-full h-full bg-white shadow-xl py-4 pb-6 flex flex-col overflow-y-auto">
+            <div class="px-4 mb-2 flex items-center justify-between">
+              <h2 class="text-lg font-medium text-gray-900">Filters</h2>
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:98]: Please localize this string.

File: app/javascript/controllers/toggle_controller.js [hunk 3]
```diff
@@ -21,12 +21,20 @@ export default class extends Controller {
       this.hasCloseClass && element.classList.remove(this.closeClasses)
       this.hasOpenClass && element.classList.add(this.openClasses)
     })
+    this.chevronTargets.forEach(element =>{
+      element.classList.remove("rotate-0")
```
Comment (joemasilotti) [inline app/javascript/controllers/toggle_controller.js:25]: If we remove `rotate-0` from the HTML then we can only worry about `rotate-180`.

Also, this class should be part of the `classes` static variable at the top of this file and set in the HTML.

File: test/integration/developers_test.rb [hunk 4]
```diff
@@ -66,12 +66,12 @@ class DevelopersTest < ActionDispatch::IntegrationTest
   test "paginating filtered developers respects the filters" do
     with_pagy_default_items(1) do
       get developers_path(sort: :availability)
-      assert_select "h2", count: 1
+      assert_select "h2", count: 2
```
Comment (joemasilotti) [inline test/integration/developers_test.rb:69]: Let's make this test more obvious by querying for an `h2` in the desktop form and an `h2` in the mobile form.

File: app/components/developer_query_component.html.erb [hunk 5]
```diff
@@ -83,8 +83,88 @@
...
+                    <%= form.collection_check_boxes(:role_types, role_types, :first, :last, {include_hidden: false}) do |b| %>
+                      <div class="flex items-center">
+                        <%= b.check_box class: "h-4 w-4 border-gray-300 rounded text-gray-600 focus:ring-gray-500", id: "mobile_role_types_#{b.object.first}", checked: role_selected?(b.object) %>
+                        <%= b.label class: "ml-3 pr-6 text-sm font-medium text-gray-900 whitespace-nowrap", for: "mobile_role_types_#{b.object.first}" %>
+                      </div>
+                    <% end %>
+                    <hr>
+                    <div class="flex items-top">
+                      <%= form.check_box :include_not_interested, class: "mt-1 h-4 w-4 border-gray-300 rounded text-gray-600 focus:ring-gray-500", include_hidden: false, checked: include_not_interested? %>
+                      <%= form.label :include_not_interested, t(".include_not_interested"), class: "ml-3 pr-6 text-sm font-medium text-gray-900" %>
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:127]: This needs the same `mobile_...` manually setting of the ID so we can tap the label to toggle the checkbox.

File: app/components/developer_query_component.html.erb [hunk 6]
```diff
@@ -83,8 +83,88 @@
               </div>
             </div>
           </div>
+    <% end %>
+    <div data-controller="toggle" data-toggle-close-class="hidden" class="relative z-10 inline-block text-left sm:hidden">
+      <div>
+        <button type="button" data-action="toggle#toggle" class="group inline-flex items-center justify-center text-sm font-medium text-gray-700 hover:text-gray-900" aria-expanded="false">
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:89]: I don't think we need the following CSS classes: `group inline-flex items-center justify-center`.

File: app/components/developer_query_component.html.erb [hunk 7]
```diff
@@ -83,8 +83,88 @@
...
+        </button>
+      </div>
+      <div data-toggle-target="element" class="bg-gray-50">
+        <div class="fixed inset-0 flex z-40 sm:hidden" role="dialog" aria-modal="true">
+          <div class="fixed inset-0 bg-black bg-opacity-25" aria-hidden="true"></div>
+          <div class="ml-auto relative max-w-xs w-full h-full bg-white shadow-xl py-4 pb-6 flex flex-col overflow-y-auto">
+            <div class="px-4 mb-2 flex items-center justify-between">
+              <h2 class="text-lg font-medium text-gray-900">Filters</h2>
+              <button type="button" data-action="toggle#toggle" class="-mr-2 w-10 h-10 bg-white p-2 rounded-md flex items-center justify-center text-gray-400 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500">
+                <span class="sr-only">Close menu</span>
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:100]: Please localize this string.

File: app/components/developer_query_component.html.erb [hunk 8]
```diff
@@ -83,8 +83,88 @@
...
+          <span><%= t(".filters") %></span>
+        </button>
+      </div>
+      <div data-toggle-target="element" class="bg-gray-50">
+        <div class="fixed inset-0 flex z-40 sm:hidden" role="dialog" aria-modal="true">
+          <div class="fixed inset-0 bg-black bg-opacity-25" aria-hidden="true"></div>
+          <div class="ml-auto relative max-w-xs w-full h-full bg-white shadow-xl py-4 pb-6 flex flex-col overflow-y-auto">
+            <div class="px-4 mb-2 flex items-center justify-between">
+              <h2 class="text-lg font-medium text-gray-900">Filters</h2>
+              <button type="button" data-action="toggle#toggle" class="-mr-2 w-10 h-10 bg-white p-2 rounded-md flex items-center justify-center text-gray-400 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-indigo-500">
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:99]:
```suggestion
              <button type="button" data-action="toggle#toggle" class="-mr-2 w-10 h-10 bg-white p-2 rounded-md flex items-center justify-center text-gray-400 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-gray-500">
```

File: app/components/developer_query_component.html.erb [hunk 9]
```diff
@@ -83,8 +83,88 @@
...
+                <span class="sr-only">Close menu</span>
+                <%= inline_svg_tag "icons/outline/x.svg", class: "h-6 w-6" %>
+              </button>
+            </div>
+            <%= form_with url: developers_path, method: :get, html: {class: "py-6"} do |form| %>
+
```
Comment (joemasilotti) [inline app/components/developer_query_component.html.erb:105]:
```suggestion
```

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking great! Amazing work on this one.

I made a few requests but they are all pretty minor. Once they are taken care of we should be good to merge.

**Comment by dhairyagabha** [comment]
@joemasilotti Ready for review

**Review by joemasilotti — CHANGES_REQUESTED** [review]
I made one small question/request regarding the hidden form field you added.

It looks like the desktop view got messed up a bit, too.

image

---

Once these are addressed we should be good to merge!

**Comment by joemasilotti** [comment]
Sounds great, thank you!

**Comment by dhairyagabha** [comment]
@joemasilotti PR Updated!
1. Mobile filters working as expected.
2. Tests added for filters in general.
3. Fixes the sorting order persistence issue.
4. Tests added for sorting.

File: app/components/sort_button_component.rb [hunk 10]
```diff
@@ -1,5 +1,5 @@
 class SortButtonComponent < ApplicationComponent
-  attr_reader :title, :name, :value
+  attr_reader :title, :name, :value, :url
```
Comment (joemasilotti) [inline app/components/sort_button_component.rb:2]: Is this change needed? I don't see `url` being used anywhere.

**Review by joemasilotti — CHANGES_REQUESTED** [review]
It looks like the sorting but is fixed for work preferences but not for timezone selection. I also made a request on an added attribute that doesn't look like it's being used.

File: app/components/sort_button_component.rb (hunk 10 above)
Comment (dhairyagabha) [reply app/components/sort_button_component.rb]: Fixed this in a recent commit, it was a residual for one of my tests. Just missed in list of changes.

**Comment by dhairyagabha** [comment]
All done.

**Review by joemasilotti — APPROVED** [review]
Great! Thanks so much for this and the back and forth. Merging in now.
